Fix deprecation warning about merging conditions

mjankowski opened the pull request:
Prior to this change there was a rails deprecation warning of:

> DEPRECATION WARNING: Merging ("accounts"."domain" IS NOT NULL) and ("accounts"."domain" = $1) no longer maintain both conditions, and will be replaced by the latter in Rails 7.0. To migrate to Rails 7.0's behavior, use `relation.merge(other, rewhere: true)`.

This occurs because the `remote` scope adds the "domain is not null" portion and the `by_domain` param when passed to `AccountFilter` leads to the domain value check.

ClearlyClaire commented:
I need to triple-check this because I remember there was a case in one of the `app/models/**_filter.rb` files where the `rewhere` behavior is actually *not* what we want.

I think this might precisely be the `remote` and `by_domain` scopes, where `remote` is not guaranteed to be processed before `by_domain` and can thus override it.

mjankowski commented:
Sounds good -- happy to refactor if there are cases this breaks.

I started down the path of adding spec coverage here but then it seemed like there already was some at the `Account` level and I wasn't sure without deeper digging which specific combinations of usage might want one behavior vs the other.

ClearlyClaire commented:
I can confirm that the parameters `by_domain` and `origin` now overwrite whichever came first, which is a behavior change. I'm not sure we can make assumptions about the ordering, especially in the API (`/api/v1/admin/accounts` and `/api/v2/accounts`), and even without this, using conflicting parameters (e.g. `by_domain=mastodon.social&origin=local`) now returns values where it should return none.

I think `AccountFilter`'s behavior should not change, but this means it'd have to be rewritten somewhat if we are to use `rewhere`.

mjankowski committed:
Add spec coverage which captures behavior of account filter

mjankowski commented:
> I think AccountFilter's behavior should not change, but this means it'd have to be rewritten somewhat if we are to use rewhere.

I agree with this, and was trying to find spec coverage which captured whatever it was supposed to be, but couldn't find any.

I just pushed one more commit which adds a few specs showing what I think the pre-rewhere-change behavior was (these specs all pass if I undo my change locally) - but which also still pass with the `rewhere` in place. Can you suggest additional cases to add here which might break after the change?

ClearlyClaire commented on spec/models/account_filter_spec.rb:
```diff
@@ -18,4 +18,29 @@
+
+  describe 'with origin and by_domain interacting' do
+    let!(:local_account) { Fabricate(:account, domain: nil) }
+    let!(:remote_account) { Fabricate(:account, domain: 'example.org') }
```
You need a second remote account with another domain to trigger test failures for `works with domain first and origin remote`

mjankowski replied:
Thanks -- will add that to specs and push an update.

ClearlyClaire replied:
One could argue this changes the behavior when `origin=remote` and `by_domain` is set to something: one would override the other while the old behavior was to return an empty list. But this probably does not matter much as that query was nonsensical in the first place.

mjankowski replied:
Yeah -- seems relatively low risk to change what was unspecified behavior before.

mjankowski committed:
Remove rewhere from AccountFilter and check params instead

In a scenario where we have both `origin=remote` and a value for
`by_domain` we can just ignore the `origin` param because its condition
will be taken care of by the domain check.

In scenarios where either origin is local or origin is remote but there
is not a domain value provided, params will be left in place as before.

mjankowski commented on app/models/account_filter.rb:
```diff
@@ -17,13 +17,13 @@ class AccountFilter
 
   def initialize(params)
-    @params = params
+    @params = params.to_h.symbolize_keys
```
The incoming params value here is sometimes a hash and sometimes a strong params object ... this change was added to normalize the strings vs symbols in the keys (some of the controllers which use the AccountFilter to some modification to the keys before using the class, and wind up with a symbols/strings mix).
